cmd/rb-main.go

```diff
@@ -0,0 +1,207 @@
+}
+
+// Validate command line arguments.
+func checkRbSyntax(ctx *cli.Context) {
+	if !ctx.Args().Present() {
+		exitCode := 1
+		cli.ShowCommandHelpAndExit(ctx, "rb", exitCode)
+	}
+}
+
+// deletes a bucket and all its contents
+func deleteBucket(url string) error {
+	targetAlias, targetURL, _ := mustExpandAlias(url)
+	clnt, pErr := newClientFromAlias(targetAlias, targetURL)
+	if pErr != nil {
+		errorIf(pErr.Trace(url), "Failed to remove `"+url+"`.")
+		return exitStatus(globalErrorExitStatus) // End of journey.
+	}
+	var isIncomplete bool
+	isRemoveBucket := true
+	contentCh := make(chan *clientContent)
+	errorCh := clnt.Remove(isIncomplete, isRemoveBucket, contentCh)
+
+	for content := range clnt.List(true, false, DirLast) {
+		if content.Err != nil {
+			errorIf(content.Err.Trace(url), "Failed to remove `"+url+"`.")
+			switch content.Err.ToGoError().(type) {
+			case PathInsufficientPermission:
+				// Ignore Permission error.
+				continue
+			}
+			close(contentCh)
+			return exitStatus(globalErrorExitStatus)
+		}
+		urlString := content.URL.Path
+
+		sent := false
+		for !sent {
+			select {
+			case contentCh <- content:
+				sent = true
+			case pErr := <-errorCh:
+				errorIf(pErr.Trace(urlString), "Failed to remove `"+urlString+"`.")
+				switch pErr.ToGoError().(type) {
+				case PathInsufficientPermission:
+					// Ignore Permission error.
+					continue
+				}
+				close(contentCh)
+				return exitStatus(globalErrorExitStatus)
+			}
+		}
+	}
+
+	close(contentCh)
+	for pErr := range errorCh {
+		errorIf(pErr.Trace(url), "Failed to remove `"+url+"`.")
+		switch pErr.ToGoError().(type) {
+		case PathInsufficientPermission:
+			// Ignore Permission error.
+			continue
+		}
+		return exitStatus(globalErrorExitStatus)
+	}
+	return nil
+}
+
+// isNamespaceRemoval returns true if alias
+// is not qualified by bucket
+func isNamespaceRemoval(url string) bool {
+	// clean path for aliases like s3/.
+	//Note: UNC path using / works properly in go 1.9.2 even though it breaks the UNC specification.
+	url = filepath.ToSlash(filepath.Clean(url))
+	// namespace removal applies only for non FS. So filter out if passed url represents a directory
+	if !isAliasURLDir(url, nil) {
+		_, path := url2Alias(url)
+		return (path == "")
+	}
+	return false
+}
+
+// mainRemoveBucket is entry point for rb command.
+func mainRemoveBucket(ctx *cli.Context) error {
+	// check 'rb' cli arguments.
+	checkRbSyntax(ctx)
+	// Set command flags from context.
+	isForce := ctx.Bool("force")
+	// Additional command speific theme customization.
+	console.SetColor("RemoveBucket", color.New(color.FgGreen, color.Bold))
+
+	var cErr error
+	for _, targetURL := range ctx.Args() {
+		// Instantiate client for URL.
+		clnt, err := newClient(targetURL)
+		if err != nil {
+			errorIf(err.Trace(targetURL), "Invalid target `"+targetURL+"`.")
+			cErr = exitStatus(globalErrorExitStatus)
+			continue
+		}
+		isEmpty := true
+		for range clnt.List(true, false, DirNone) {
```

Comment:
I think you need to pass DirFirst here, we can have some empty directories sometimes

Reply:
@vadmeste, it is my understanding that DirFirst option would allow listing directory entries ahead of files. Here, the purpose is to ascertain if the bucket in targeturl is empty , hence listing with DirNone. fwiw, I tested with empty directory and it behaves correctly.
```
➜  minio git:(master) mc ls myminio/newb --r
[2019-02-13 23:01:58 PST]     0B em3/
➜  minio git:(master) mc mb myminio/newb    
Bucket created successfully `myminio/newb`.
➜  minio git:(master) mc ls myminio/newb --r    
[2019-02-13 23:03:47 PST]     0B em3/
➜  minio git:(master) mc rb myminio/newb
mc: <ERROR> `myminio/newb` is not empty. Retry this command with ‘--force’ flag if you want to remove `myminio/newb` and all its contents 
➜  minio git:(master) mc rb myminio/newb --force
Bucket removed successfully `myminio/newb`.
```